sqlserver/datadog_checks/sqlserver/schemas.py

```diff
@@ -0,0 +1,369 @@
+
+    def set_base_event_data(self, hostname, tags, cloud_metadata, dbms_version):
+        self._base_event["host"] = hostname
+        self._base_event["tags"] = tags
+        self._base_event["cloud_metadata"] = cloud_metadata
+        self._base_event["dbms_version"] = dbms_version
+
+    def reset(self):
+        self._columns_count = 0
+        self._total_columns_count = 0
+        self.db_to_schemas = {}
+        self.db_info = {}
+
+    def store_db_info(self, db_name, db_info):
+        self.db_info[db_name] = db_info
+
+    def store(self, db_name, schema, tables, columns_count):
+        self._columns_count += columns_count
+        self._total_columns_count += columns_count
+        schemas = self.db_to_schemas.setdefault(db_name, {})
+        if schema["id"] in schemas:
+            known_tables = schemas[schema["id"]].setdefault("tables", [])
+            known_tables = known_tables + tables
+        else:
+            schemas[schema["id"]] = copy.deepcopy(schema)
+            schemas[schema["id"]]["tables"] = tables
+        if self._columns_count > self.MAX_COLUMN_COUNT:
+            self._submit()
+
+    def exceeded_total_columns_number(self):
+        return self._total_columns_count > self.MAX_TOTAL_COLUMN_COUNT
+
+    #NOTE: DB with no schemas is never submitted
+    def submit(self):
+        if not bool(self.db_to_schemas):
+            return
+        self._columns_count = 0
+        event = {**self._base_event, "metadata": [], "timestamp": time.time() * 1000}
+        for db, schemas_by_id in self.db_to_schemas.items():
+            db_info = {}
+            if db not in self.db_info:
+                self._log.error("Couldn't find database info for %s", db)
+                db_info["name"] = db
+            else:
+                db_info = self.db_info[db]
+            event["metadata"] = event["metadata"] + [{**(db_info), "schemas": list(schemas_by_id.values())}]
+        json_event = json.dumps(event, default=default_json_event_encoding)
+        self._log.debug("Reporting the following payload for schema collection: {}".format(json_event))
+        self._submit_to_agent_queue(json_event)
+        self.db_to_schemas = {}
+
+
+def agent_check_getter(self):
+    return self._check
+
+
+class Schemas:
+
+    # Requests for infromation about tables are done for a certain amount of tables at the time
+    # This number of tables doesnt slow down performance by much (15% compared to 500 tables)
+    # but allows the queue to be stable.
+    TABLES_CHUNK_SIZE = 50
+
+    def __init__(self, check, schemas_collection_interval):
+        self._check = check
+        self._log = check.log
+        self.schemas_per_db = {}
+
+        base_event = {
+            "host": None,
+            "agent_version": datadog_agent.get_version(),
+            "dbms": "sqlserver",
+            "kind": "sqlserver_databases",
+            "collection_interval": schemas_collection_interval,
+            "dbms_version": None,
```

Comment:
`dbms_version` can be grabbed via https://github.com/DataDog/integrations-core/blob/master/sqlserver/datadog_checks/sqlserver/statements.py#L424

Reply:
mm it's not None actually, it gets set along with host in set_base_event_data. 

host - may change. 

version -  probably not ? I did it this way cause version is fetched at the end of the SqlServer check constructor and is unknown when Schema class is created. Can be lazily initialised though on the first check.